### PR TITLE
introduce dstringt::starts_with

### PR DESCRIPTION
**Pull request by kroening**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by NlightNFotis — COMMENTED** [review]
There's a weird build failure on Travis, on osx, but aside from that, this is good stuff!

I very much like the readability of this compared to `has_prefix`. Does it mean that `has_prefix` is now deprecated? If yes, should it be marked so?

**Comment by hannes-steffenhagen-diffblue** [comment]
Does that mean we no longer care about being able to swap in `std::string` for `dstringt`? No objection to that personally, but if we do that we should also remove `USE_STD_STRING` (that’d have to happen before this PR because we do actually test if cbmc builds with `USE_STD_STRING`).

**Comment by tautschnig** [comment]
> Does that mean we no longer care about being able to swap in `std::string` for `dstringt`? No objection to that personally, but if we do that we should also remove `USE_STD_STRING` (that’d have to happen before this PR because we do actually test if cbmc builds with `USE_STD_STRING`).

I would believe we maintain full compatibility: with this patch, `.starts_with()` can be used the same way with `dstringt` and `std::string`.

@kroening I'd +1 the suggestion to deprecate `has_prefix`, and in a lot of files the `#include <util/prefix.h>` should be removed.

**Comment by kroening** [comment]
We can either wait for C++20 (which adds `.starts_with()` to `std::string`) or remove `USE_STD_STRING`.

**Comment by kroening** [comment]
I've dropped the second commit until we do C++20. The first commit is useful in preparation.

File: src/util/dstring.h [hunk 1]
```diff
@@ -90,6 +90,22 @@ class dstringt final
     return no==0; // string 0 is exactly the empty string
   }
 
+  /// equivalent of as_string().starts_with(s)
+  bool starts_with(const char *s)
+  {
+    for(const char *t = c_str(); *s != 0; s++, t++)
+      if(*t != *s)
```
Comment (tautschnig) [inline src/util/dstring.h:97]: This is out-of-bounds when `this->size()` is less than `strlen(s)`.

**Review by tautschnig — CHANGES_REQUESTED** [review]

File: src/util/dstring.h (hunk 1 above)
Comment (kroening) [reply src/util/dstring.h:97]: `c_str` guarantees zero-termination.

**Review by tautschnig — APPROVED** [review]

**Pull request merged**